### ISSUE TITLE
implementing groups

### ISSUE THREAD

**Issue by zeffii** [nortikin/sverchok#785]
For my own sanity i felt it was necessary to have a closer look at the mindfuck that is CustomGroup Nodes :) . I was especially interested in reusing as much of the PyNode api as possible, but I think @ly29 already concluded that also this requires us to pretty much _roll our own_.

I started a branch https://github.com/nortikin/sverchok/tree/group_experiment which hopes to get away from the use of the Frame and instead be more like the ShaderNodeTree with a back button and an overlay to show the parent NodeTree.

------------
### This is merely an exploration i'm not married to any of this code

here a list of things that need to be taken care of (feel free to interject / edit) in no order of importance.

- [x] socket rename  
- [x] deleting existing sockets  
- [x] moving existing sockets  
- [x] socket type recast
- [x] handle reroutes  
- [x] generate group from selected 
- [x] generate group from selected + auto connect head and tail to monad IO, and parent node appropriately    
- [ ] sort relinks by Y location of periphery (aesthetic - avoid crossing noodles) 
- [ ] show relevant properties in the sockets of the node tree  
- [ ] handle node group copy  
- [ ] handle node group rename
- [ ] expand monad to ordinary nodes  
- [ ] handle IO/json  
- [ ] handle script nodes  
- [ ] possible `space_data.node_tree`  (the `[ nodetree icon ][ +  New   ][pin]`  ) should show the parent node tree name as in ShaderNodeTree ..
- [ ] if possible keep old node BETA groups working.

**Comment by zeffii** [comment]
I don't know if i'll have time to do anythig with this, I hope to reuse some of @ly29 's code (which looks fine) seen as it appears we'll really need to code all the bits, and then add retrigger updates for they node group dependencies.

**Comment by ly29** [comment]
Yeah. my code ran in to issues with all the possibilities of editing. 
The UI issues also became annoying.
If anything seems off in that code it probably is.

**Comment by zeffii** [comment]
i'll take a long stare at it, and if possible implement something we can elaborate on. I'm not yet convinced   that it's a dead :)

**Comment by ly29** [comment]
I don't think it is unworkable. I just wanted to do the Redux project, for which I later didn't have time. Sadly, but I still want scratch that itch.

**Comment by ly29** [comment]
That sad this is a small scale project which I might have the time to play around with a bit.

**Comment by zeffii** [comment]
yeah, understanding the plumbing of this can be useful for redux or whatever :)

**Comment by zeffii** [comment]
something weird is happening.. in `space_node.py`

```python

class NODE_UL_interface_sockets(bpy.types.UIList):
    def draw_item(self, context, layout, data, item, icon, active_data, active_propname, index):
        socket = item
        color = socket.draw_color(context)

```

I think `draw_item` is called for each new socket generated by the Group Input / Output nodes. Surprisingly `socket` doesn't have a `draw_color` attribute. In our `node_tree.py` it shows our sockets do all have this attribute, so perhaps some information is being stripped?

**Comment by zeffii** [comment]
printing the socket info..
```
            print(dir(socket))
            print(socket.bl_socket_idname)
            print(socket.identifier)
            print(socket.name)
            print(socket.rna_type)

['__doc__', '__module__', '__slots__', 'bl_rna', 'bl_socket_idname', 'identifier', 'is_output', 'name', 'rna_type']
StringsSocket
Input_1
Start
<bpy_struct, Struct("NodeSocketInterface")>
```

shows this is indeed the case...

**Comment by zeffii** [comment]
It looks like the `Group Input` node is both a node and a part of   `this_nodegroup.inputs`,  it seems that the `sv_node_group_in` would need to inherit from multiple places... arghhh! :)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17171353/a2feb504-53f0-11e6-8161-82e66f0ba08b.png)

but only by modifying node_tree.py, which is not an option for sverchok..

```python
class NODE_UL_interface_sockets(bpy.types.UIList):
    def draw_item(self, context, layout, data, item, icon, active_data, active_propname, index):
        socket = item
        
        if hasattr(socket, 'draw_color'):
            color = socket.draw_color(context)
        else:
            color = {
                "StringsSocket": (0.6, 1.0, 0.6, 1.0),
                "VerticesSocket": (0.9, 0.6, 0.2, 1.0),
                "MatrixSocket":  (.2, .8, .8, 1.0)
            }.get(socket.bl_socket_idname, (0.2, 0.2, 0.2, 1.0))
```

**Comment by ly29** [comment]
Monkeypatch is a solution but not a good one. Node in and out needs to be redone (obviously)

**Comment by zeffii** [comment]
Yeah, i'm only modifying `space_node.py` to get an appreciation for what's going on.

**Comment by ly29** [comment]
A god way to understand the context Sverchok is working in.

**Comment by ly29** [comment]
Okay I have looked a my own code a bit as well as yours some small issues to resolve pops up.

Where to store the sv groups, I choose a specific node group tree type to mimic the cycles behavior where the nodes groups are hidden and keep the actual editing via the interface only.

Also looking at  ```space_node.py``` the socket used there are of type [NodeSocketInterface](https://www.blender.org/api/blender_python_api_2_77_1/bpy.types.NodeSocketInterface.html?highlight=socketinterface#bpy.types.NodeSocketInterface) which has ```draw_color(self, context)``` but ordinary [NodeSocket](https://www.blender.org/api/blender_python_api_2_77_1/bpy.types.NodeSocket.html) has ```draw_color(self, context, node)```. The reason for this change of signature escapes me right now between the different classes escape me right now.



Also, some comments never hurts, took a while to understand my thinking in ```group.py```

**Comment by ly29** [comment]
Basically the behavior of editing needs to be well defined.

My implementation uses the [```SverchGroupTree```](https://www.blender.org/api/blender_python_api_2_77_1/bpy.types.NodeSocket.html) tree type

Looking back I am quite happy with my mixin class ```StoreSockets```, almost a bit smug. Needs some tweaking of course.

Some other things I did makes me cringe, for example the ```sv_init``` change is less clear to me now. Also how node groups actually runs and how they copy data back and forth seems less than ideal.

**Comment by ly29** [comment]
So we need to find out how our sockets can be converted into interface sockets...

**Comment by ly29** [comment]
I guess we would have to create matching ```NodeSocketInterface``` classes like ```NodeSocketInterfaceVectorVelocity``` etc.

How does the matching mechanism for creating ```NodeSocketInterfaceInt``` from ```NodeSocketInt``` look?

**Comment by zeffii** [comment]
>  So we need to find out how our sockets can be converted into interface sockets...   

this . and there's no documentation that i'm aware of. not even hints... just the C code if you can make sense of it.

> Where to store the sv groups, I choose a specific node group tree type to mimic the cycles behavior where the nodes groups are hidden and keep the actual editing via the interface only.

yeah, not cluttering the node tree list with groups. Tho this wasn't a concern while exploring.

-----
There's a good reason to ask LukasToenne about this, because we're getting close enough to "know what we don't know" :)  and if there was a way to reuse some of the existing pynodes NodeSocket interface stuff that would make our life easier and anyone interested in similar.


> I guess we would have to create matching NodeSocketInterface classes like NodeSocketInterfaceVectorVelocity etc.
> How does the matching mechanism for creating NodeSocketInterfaceInt from NodeSocketInt look?

I haven't spent a lot of time trying to figure this out, but blender github repo by dfelinto allows for [convenient online searching](https://github.com/dfelinto/blender/blob/3068ea34e414e7af6d20af7ab92c69cbd9a33414/source/blender/blenkernel/intern/node.c#L595-L624)

**Comment by zeffii** [comment]
I think even if the `NodeCustomGroup`  can't be usefully subclassed that the concepts you offered in `group.py` can be extended to replicate all features

... just a bit of heartburn to have to re-implement everything.

**Comment by ly29** [comment]
Yes, I vaguely remember that it might not be finished, however it is worth asking just keep it in the todo or work out a solution. 
And I might very well be wrong. 
I looked into the c code but couldn't quite find the magic.

**Comment by zeffii** [comment]
I think at least one practical limitation to our own implementation of the interface panel is that it would be a bit of a pain to draw the sockets in the UI lists with the correct colors..  a cosmetic wart. Maybe settle for a single letter indicator or a small FloatVectorProperty subtype=RGBA

**Comment by zeffii** [comment]
when ShaderNodeTree nodes are added to a Custom group, upon inspection like:

```python
bpy.data.node_groups['NodeGroup'].inputs[0].rna_type.name
>>> 'Color Node Socket Interface'
```

but when I add a sv node / socket to a nodegroup input, it just says    

```python
>>> 'Node Socket Template'
```

**Comment by zeffii** [comment]
so maybe we have to overwrite the rna_type of out custom sockets?

**Comment by zeffii** [comment]
haha .. i have no idea about any of this.

**Comment by ly29** [comment]
Neither do I, based on context we could create the right type of sockets. But it does seem somewhat backward. 

Asking Lukas seems better...

**Comment by zeffii** [comment]
well some poking around, might be not super useful.. but poking anyway:
```python
>>> for _ in list(bpy.data.node_groups['rtt'].inputs[0].rna_type.properties):
...     print(_)
...     
<bpy_struct, PointerProperty("rna_type")>
<bpy_struct, StringProperty("name")>
<bpy_struct, StringProperty("identifier")>
<bpy_struct, BoolProperty("is_output")>
<bpy_struct, StringProperty("bl_socket_idname")>

>>> for _ in list(bpy.data.node_groups['rtt'].inputs[0].rna_type.properties):
...     print(_)
...     print(_.rna_type)
...     print(_.name)
...     print(_.identifier)
...     print(_.is_output)
...     print(_.bl_socket_idname)
...     
<bpy_struct, PointerProperty("rna_type")>
<bpy_struct, Struct("PointerProperty")>
RNA
rna_type
False

Traceback (most recent call last):
  File "<blender_console>", line 7, in <module>
AttributeError: 'PointerProperty' object has no attribute 'bl_socket_idname'

>>> 
```

**Comment by zeffii** [comment]
```python

for p in list(bpy.data.node_groups['rtt'].inputs[0].rna_type.properties):
    for attr in dir(p):
        print(attr, ':', getattr(p, attr))

__doc__ : None
__module__ : bpy.types
__slots__ : ()
bl_rna : <bpy_struct, Struct("StringProperty")>
default :
description :
icon : NONE
identifier : bl_socket_idname
is_animatable : False
is_argument_optional : False
is_enum_flag : False
is_hidden : False
is_library_editable : False
is_never_none : True
is_output : False
is_readonly : False
is_registered : True
is_registered_optional : True
is_required : False
is_runtime : False
is_skip_save : False
length_max : 0
name : ID Name
rna_type : <bpy_struct, Struct("StringProperty")>
srna : None
subtype : NONE
translation_context : *
type : STRING
unit : NONE
```

**Comment by ly29** [comment]
So the ```NodeSocketInterface*``` are the input and outputs for the _NodeGroup_, when we create the node group we could set them I guess.
```
ng.inputs.new("StringsSocketInterface", "x'")
```
It is starting to make sense now perhaps.

**Comment by zeffii** [comment]
lukas_t made his own node group for object_nodes: 
https://developer.blender.org/diffusion/B/browse/object_nodes/release/scripts/nodes/group_nodes.py

**Comment by ly29** [comment]
nice to have a reference

**Comment by zeffii** [comment]
nice use of dynamic class's `type()` :)

**Comment by ly29** [comment]
As a gist for easier viewing for me right now.
https://gist.github.com/ly29/fc31153e9a0fecdd5703d0306d40bb31

**Comment by ly29** [comment]
Now I think we have all the knowledge we need. It looks fairly clear now. Sure a lot of issues will crop up...

I will be away over the weekend.

**Comment by zeffii** [comment]
okidoki... i'm going to mess around with this..

![image](https://cloud.githubusercontent.com/assets/619340/17257797/019c5d4e-55c4-11e6-8a93-090a43f128fd.png)

**Comment by zeffii** [comment]
@ly29  copied you process function from the group node almost verbatim :)

![image](https://cloud.githubusercontent.com/assets/619340/17263173/da81aaf4-55df-11e6-8e80-1fac1d55e882.png)

**Comment by ly29** [comment]
Very nice looking. 
That '''process''' is almost sane even though it kind of breaks the separation of where thing should be happening...

**Comment by zeffii** [comment]
I do like your approach to having the interface directly on the Input and Output nodes, instead of the Panel, but I wanted to see the least amount of code (not withstanding the fact that the process of In and Out are virtually the same except for  `self.inputs` and `self.outputs` and using `links[0].from_socket` vs `links[0].to_socket` ) i felt it was too early to write dynamic code.. but maybe it makes sense to do exactly that already.

**Comment by ly29** [comment]
The group inputs don't need process, but update because it only dealing with UI issues.

**Comment by zeffii** [comment]
here a list of things that need to be taken care of (feel free to interject / edit) in no order of importance.

- deal w node group copy
- deal w node group rename
- work w IO/json
- work w script nodes
- moving existing sockets
- deleting existing sockets
- possible `space_data.node_tree`  (the `[ nodetree icon ][ +  New   ][pin]`  ) should show the parent node tree name as in ShaderNodeTree ..

**Comment by zeffii** [comment]
Oh i see what you mean now, yeah process i'll move back into the parent Node.. it may even solve the `red` color of the nodes inside the group..

**Comment by ly29** [comment]
They shouldn't be running while being edited I think. Mmm. 

Also the issue of groups within groups. Sanity checks to avoid infinite recursion could be a good idea even though it should the max recursion depth pretty quickly in most cases...

- Also if they should be hidden like with cycles, if we have two types of node layouts - one for groups, one regular like my earlier approach
- making avilable groups visible via shift-a menu
- also think about if there is a need to limit certain nodes from being in a group, mostly nodes with side effects should be kept out of groups. However not certain that is limit we should impose, perhaps encourage

**Comment by ly29** [comment]
Having the editing directly on the node feels logical even if we have make it a bit bigger.

**Comment by zeffii** [comment]
maybe we can call these node_groups something unique to help disambiguate `bpy.data_node_groups`. -- i was thinking `sub_groups` or `nested_groups`. (altho the distinction may be unimportant to non scripters)

yeah a superset seems logical like non of the view nodes or wifi.

regarding the I/O nodes, we can either set a default width to show all operators nicely, or make the UI respond to width resizes and display only sockets if it's minimally wide and when resized it can display the move / delete operators. no haste with that :)

**Comment by ly29** [comment]
Node groups is an bad name that I guess we are stuck with now. 
How about node functions?
I always think about them as function and also brings across the side effects clearly. 
Wifi within a group could be fine, but better think about restricting them. 
Also as much I dislike wifi cross layout would be nice. However that has the side effect of imposing an order among different layouts. (For later anyway)
But restrictions are more of usability issue which isn't urgent.

**Comment by enzyme69** [comment]
How about Node Sverchok Sub-Network? Is this for "grouping of bunch of
network to reuse as a single node"?



------------------------------------------------------------
- Jimmy Gunawan
http://blendersushi.blogspot.com.au
http://mayaspiral.blogspot.com.au
http://houdoodles.blogspot.com.au
http://puppetar.blogspot.com.au

**Comment by zeffii** [comment]
yeah anything but node_groups.   
- `functors`: ( i like it because it's short), 
- `functions`: ( a little broad, but i don't oppose it)
- `node_functions`: yeah. works for me.
- 'monads' :  (compositional units)
- 'asbtract_nodes`: they are abstractions afterall from composed nodes..

**Comment by zeffii** [comment]
Hi @enzyme69 I'd like to keep the names short to help us talk about them easier.

**Comment by zeffii** [comment]
there seems to be a problem with duplicate socket names,

**Comment by ly29** [comment]
Use socket.identifier instead :)

**Comment by ly29** [comment]
Or index

**Comment by zeffii** [comment]
what I should have said was:

while it is possible to connect multiple sockets with the same name to the group's inputs, they will be added just fine, but this will fail..
```python
        for socket in self.inputs:
            if socket.is_linked:
                data = socket.sv_get(deepcopy=False)
                in_node.outputs[socket.name].sv_set(data)
```

so yeah we can use index there, probably makes sense. 

but just like with functions we shouldn't really be allowing multiple default keyworded parameters to have the same name like `def some_function(x=x, y=y, x=x):`

**Comment by zeffii** [comment]
or maybe we shouldn't limit same name sockets..  i can imagine one use positive usecase

```python
[_____ node group_____  ]
o     shape 1
o     radius
o     height
o     shape 2
o     radius
o     height
```

where there is context

**Comment by ly29** [comment]
I agree, I think it is a somewhat strange design decision, but the names of sockets are more like labels than the name function arguments. 
For Sverchok I think there are other places it will fail if the socket names aren't unique

**Comment by ly29** [comment]
Data lookup should work since it uses the socket.identifier so. It might break somewhere else but I think it should work.

**Comment by zeffii** [comment]
> For Sverchok I think there are other places it will fail if the socket names aren't unique

i don't think we were lazy if that's the case, it makes sense to be consistent with uniqueness.

i'm using `socket.index` at the moment, will change if it's obviously hurting us.

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17269838/51676a40-5654-11e6-803c-86e7e563e394.png)

**Comment by zeffii** [comment]
i think _renaming sockets_ is top of my list now.

**Comment by ly29** [comment]
Organizing the node groups and naming them is very useful. 
Also worth looking into the possibility of auto migrating old node groups to new ones.

**Comment by zeffii** [comment]
will need to learn about UIList :) I think this is exactly what we want..
[index-out-of-range-for-uilist-causes-panel-crash/14203#14203](http://blender.stackexchange.com/questions/14202/)

**Comment by ly29** [comment]
The reference document is quite good
https://www.blender.org/api/blender_python_api_2_70_5/bpy.types.UIList.html

**Comment by zeffii** [comment]
We won't need both, some decision has to be made between the interface directly on the I/O nodes, or in the panel. I'm starting to lean even more towards directly on the node UI as with your original implementation.

**Comment by zeffii** [comment]
hmmm, but this doesn't seem so heavy : https://gist.github.com/zeffii/cf86ac59f954403ab81b81f428f84e3a#file-node_groups-py-L123-L179

the rest of the code does seem quite elaborate, like a tank.... (but heavy?... )

**Comment by ly29** [comment]
We can even try both.

**Comment by ly29** [comment]
I really like editing directly on the node, going to the panel feels a bit complicated, however it is the blender way...

**Comment by zeffii** [comment]
The blender way isn't always the best. Am tempted to get as much of your node_group socket wrappers understood and implemented.. on the other hand I would like to try my own version of the panel (even if it doesn't make it into a final/accepted version).

**Comment by ly29** [comment]
As my design teachers used to say, don't be afraid to try! And don't be afraid to change and do things over again. 

About the blender way, as much as I like blender sometimes it does drive me mad. But that is true for most advanced software...

**Comment by zeffii** [comment]
trying to keep it simple...
![image](https://cloud.githubusercontent.com/assets/619340/17275601/c9647f74-570d-11e6-8d28-d3cac0ee66a3.png)

**Comment by zeffii** [comment]
each is an operator, which will be able to edit the Input node and the parent NodeGroup interface..  probably implementing this will lead me to want to work harder on understand the UiList approach, even tho at the moment that seems monsterously complex....

**Comment by ly29** [comment]
I itch a bit to play around with as well right now. UIList would be good for node tree manager as well.

**Comment by ly29** [comment]
Looking nice brw

**Comment by zeffii** [comment]
ok renames work.. but wow is it fugly code...

**Comment by zeffii** [comment]
ha, I have no shame :)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17275949/ec54776c-5719-11e6-9c1d-64f07da05928.png)

**Comment by zeffii** [comment]
with these kinds of situations, might have to use `.prop_name' instead..
```
        self.inputs.new('StringsSocket', "Step").prop_name = 'stop_'
        self.inputs.new('StringsSocket', "Stop").prop_name = 'step_'
```

**Comment by zeffii** [comment]
and i suppose this is where hidden sockets will mess with the `sockets.move(from_index, new_index)` unless guarded against.. haha .. headfuck

**Comment by ly29** [comment]
The whole issue of having the properties in the nodes comes back to bite us. 
Of course we could make dynamic nodes with type and copying the custom properties. Or something. 

The other option is to expose the relevant properties. 
But that includes creating an extra node and managing all of that...
The best long term option given current blender limits is making a hierarchy of sockets like the standard blender ones...

**Comment by zeffii** [comment]
renaming is complete.  :)  onto delete :)

**Comment by zeffii** [comment]
the move / delete / rename / type switching operators share a bit too much code ..not to urge me to turn in into one callback operator or something.. (eventually).

**Comment by zeffii** [comment]
hmm, this is a bit less of a headfuck than I thought it would be.  the hidden sockets on existing nodes don't really get in our way, seen as i'm only using indices, and we don't move/remove sockets on nodes that aren't part of the node_group ecosystem (parent node and I/O nodes)

**Comment by zeffii** [comment]
cool. moving sockets is easy now :)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17277393/0af87614-5742-11e6-8656-30e9ffe6e0ec.png)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17277624/b1b1b8a8-5747-11e6-9520-ae296a42f235.png)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17277775/c1fee4fc-574b-11e6-8181-1b14377391c9.png)

**Comment by zeffii** [comment]
![responsive_ui](https://cloud.githubusercontent.com/assets/619340/17279005/f31ae030-576a-11e6-9b7f-978ec74f2b66.gif)
responsive UI

**Comment by ly29** [comment]
Been playing around and some observations of issues, not necessarily in order of importance or directly urgent.

- one node function (group) can be present in many places, editing should update all and or provide a way to make unique. this is maybe a little bit tricky 

- it does not handle reroutes, to get a corresponding socket there is ```get_other_socket``` in data_structure, now extended to handle down and up (this cycles node groups doesn't support) but it is simple and makes sense.

**Comment by zeffii** [comment]
good points.

- The copy / make unique is next up:
  - I think the simplest approach would be an OK preliminary solution, that is to say: consider the act of duplicating a node as synonymous to _make unique copy_.
  - The more complex approach (and likely going to slow us down) would be to treat a _sub group_ as something that can be compiled and turned into a proper function (composed visually as nodes, but represented as code inside the parent node.: something like:
        ```python
        # add functions
        range_float_001 = RangeFloat()
        math_001 = MathNum()
        math_002 = MathNum()

        # set function props
        range_float_001.mode = 'COUNT'
        range_float_001.stop = tau
        math_001.mode = 'SIN'
        math_002.mode = 'COS'

        # make function connections
        inputs[0] > range_float_001.inputs[2]
        range_float_001.outputs[0] > math_001.inputs[0]
        range_float_001.outputs[0] > math_002.inputs[0]

        # etc...        
         ```
- The reroute I haven't looked (probably because I don't use it much... easy to overlook the smaller but cool things)

**Comment by ly29** [comment]
Compiling would slow us down a lot, right now the code in many cases assumes that the ```process``` function lives in blender node group etc, now reason to change that right now. for redux it could be a possibility.

I don't agree that copying the node should make the node group unique. Every node group simple has a series of parent nodes and every node group node has one node group.

Something like this in node tree (this case assume a specific node group type for storing the node trees).
```
    @property
    def parent_nodes(self):
        parents = []
        for ng in filter(bpy.data.node_groups, lambda ng: ng.bl_idname == "SverchCustomTreeType"):
            for node in ng.nodes:
                if node.bl_idname == "SvGroupNode" and node.group_name == self.name:
                    parents.append(node)
        return parents
```

Then we update them all, otherwise the functions/monads aren't reusable

**Comment by zeffii** [comment]
> I don't agree that copying the node should make the node group unique.

I didn't mean to suggest that, copying ideally should offer to ability to 
   1. make unique
   2. make duplicate ( edit one, edit all )

**Comment by ly29** [comment]
Okay, good, my mistake. 😄 

Anyway, a decision that should be taken soon is if we hide away the node groups in an other node tree type that makes them invisible in the selection.

Also, just for fun, you can edit the old SvGroup implementation if you enter the right node group name in Group Node and types edit...
![skarmavbild 2016-08-01 kl 10 29 25](https://cloud.githubusercontent.com/assets/6241382/17287969/e42727e2-57d2-11e6-8bb4-9cb1dac0d11a.png)

**Comment by ly29** [comment]
There is a lot of duplication going around, the user interface and your code mostly looks cleaner than mine old, but if we modify my old previously created node groups will still work.

There are some assumptions about editing in my old code and it does not seem to work very well.

**Comment by zeffii** [comment]
I attempted to as little hard-coding as possible. The creation of `SverchCustomTreeType` was the exception. I'm all in favour of switching it to SverchGroupTree. My code also makes a few assumptions about the `space_data.path` (ie it must be have 2 entires). but it uses a custom IDProperty ['sub_group'] to determine if it should display the interface ( I consider these temporary or semi permanent until there's an obviously better way to do it)

**Comment by zeffii** [comment]
`SvGroupEdit` operator sort of prevents two groups conflicting in name:

```python
    def execute(self, context):
        node = context.node
        parent_tree_name = node.id_data.name
        ng = bpy.data.node_groups

        group_node = ng.get(self.group_name)
        if not group_node:
            group_node = group_make(node, new_group_name=self.group_name)
```
something could be done here..

**Comment by ly29** [comment]
Regarding renamning, isn't it the new one that gets the .001 name?

```
>>> D.node_groups.new("MyNodeTree", "SverchCustomTreeType")
bpy.data.node_groups['MyNodeTree']

>>> ng  = D.node_groups.new("MyNodeTree", "SverchCustomTreeType")
>>> ng.name
'MyNodeTree.001'
```

**Comment by zeffii** [comment]
I hope so :) that being the case that does make things a little less tricky (still juggling tho)

**Comment by zeffii** [comment]
![gif](https://cloud.githubusercontent.com/assets/619340/17290241/c7b70432-57dd-11e6-8a50-3fa8098d3440.gif)

**Comment by zeffii** [comment]
I just read Lukas_t's Interface code, and I think I understand it now after implementing my own.. neat to see much of the same things happening in both implementations..  I kind of like my approach, but might give his approach a run anyway..

**Comment by ly29** [comment]
If nothing else you learn something about a different style. 
I think I going to give my old group creation from selection code a test

**Comment by zeffii** [comment]
I had hoped to propagate the prop_name somehow, but i get this weird error:

    `rna_uiItemR: property not found: SvGroupNodeExp.stop_`

```python
            if not use_socket_prop_name:
                getattr(parent_node, _puts).new(new_type, new_name)
            else:
                f = getattr(parent_node, _puts).new(new_type, new_name)
                f.prop_name = socket_prop_name
```

**Comment by zeffii** [comment]
oooh... i remember you had to do some magic for this dynamic stuff to work in Scripted node.... echh...i think i'll pass for now.

**Comment by zeffii** [comment]
well.. not magic... but definitely tapdancing around.

**Comment by ly29** [comment]
The sane approach, in retrospect, would be to have a standard hierarchy of sockets with the different default values, even if we would need more than we cycles.
having the properties in the node wasn't clever, but at the time we didn't know this problem.

now if we want to have the proper props display we need to replicate all of this, or create generic display mechanism like script did.

for me the most sane approach would be to create node representing the node group interface _somwwhere_ and then expose the property when needed in the node interface.

therefore I conclude that sockets are a mess the way they are done Sverchok today.

**Comment by zeffii** [comment]
the bpy.props.IntProperty  or FloatProperty would come from the Socket,  I think only the `StringsSocket` has any real usefulness as a _on-ui-slider_.

**Comment by ly29** [comment]
Yeah, but it would be nice with different types, locked between say 0 to 1 as the original nodes do.

**Comment by zeffii** [comment]
arbitrary post-init min/max would be great yeah..

**Comment by zeffii** [comment]
ok managed to get some UiList to populate the Panel with the info from `SvGroupInputsNodeExp.outsputs`

**Comment by zeffii** [comment]
i'm just going to call node_groups/sub_groups "monads", you'll know what i'm talking about.

The weird/klunky thing (to me) about using a UiList to display + edit the props is that each monad must necessarily have a unique pair (tied to the group, but not necessarily to groupname) registered at monad instantiation. 

I have questions, well one main thing..
- would F8 leave such things untouched, ie don't need to re-register

**Comment by zeffii** [comment]
My apprehension to going with UILists grows either due to laziness or knowing that in order for UILists to work we might as well copy most of Lukas' object_node node group implementation.

I'm not sure there's a net gain either way, but probably implementing UILists for the Interface should happen in a separate branch, to keep that implementation closer to the example.

**Comment by zeffii** [comment]
dynamically create properties:  http://blender.stackexchange.com/a/1786/47 have not tried yet, but seems interesting...

**Comment by ly29** [comment]
To get properties to show without having to invent a whole new socket system there are some options I can think of, listed in how "clean" they make me feel.

- Create a dynamic node class with ```type```  for each monad containing the relevant properties, that we get the options for by parsing the nodes
- Create a node for each property in the _some_ layout (another node category) and modifying socket to be able to display that property, also some simple modifications to the update system are needed 
- A script node like option without the control of the properties

Somehow the second option kind of appeals me right now.

Some notes:

For option one, it is trivial to collect the needed info for the properties for the dynamic class creation.

```
>>> ng.nodes['Sphere'].inputs['Radius'].prop_name
'rad_'

>>> ng.nodes['Sphere'].rna_type.rad_[0]
<built-in function FloatProperty>

>>> ng.nodes['Sphere'].rna_type.rad_[1]
{'options': {'ANIMATABLE'}, 'description': 'Radius', 'update': <function updateNode at 0x116560268>, 'name': 'Radius', 'attr': 'rad_', 'default': 1.0}
```

For two the question is where to store them, the answer would be, I guess, in the node monad layout when we save it or another node tree type which shadows the node groups by name and contains relevant properties. this is backward and complicated, but pure brutish hack of it seems kind of fun. This also has the advantage that more complicated things might work better.

The script node method of having a number of fixed properties and using them feel ok there but here it feels out of place since the properties won't be well behaved.

**Comment by zeffii** [comment]
What about an `SvInterfaceSocketType` which has like.. 10 standard property offerings..  none of this is i've tried .. but am tempted to..
```python
class SvInterfaceSocket(NodeSocketStandard):
    '''String any type - one string'''
    bl_idname = "SvInterfaceSocket"
    bl_label = "Interface Socket"

    dynamic_prop_name = StringProperty(default='')
    dynamic_prop_type = StringProperty()  # or Enum...
    dyn_bool_limit_0_1 = BoolProperty()
    dyn_int_limit_0_10 = IntProperty(min=0, max=10)
    dyn_int_limit_0_100 = IntProperty(min=0, max=100)
    dyn_int_limit_1_1 = IntProperty(min=-1, max=1)
    dyn_int_limit_10_10 = IntProperty(min=-10, max=10)
    dyn_int_limit_100_100 = IntProperty(min=-100, max=100)

    def sv_get(self, default=sentinel, deepcopy=True):
        if self.is_linked and not self.is_output:
            return SvGetSocket(self, deepcopy)
        elif self.dynamic_prop_name:
            return [[getattr(self.node, self.prop_name)]]
        elif default is not sentinel:
            return default
        else:
            raise SvNoDataError

    def sv_set(self, data):
        SvSetSocket(self, data)

    def draw(self, context, layout, node, text):
        if self.dyn_prop_name:
            if self.is_output:
                t = text
                msg = "Warning output socket: {name} in node: {node} has property attached"
                print(msg.format(name=self.name, node=node.name))
            else:
                prop = node.rna_type.properties.get(self.prop_name, None)
                t = prop.name if prop else text
        else:
            t = text

        if not self.is_output and not self.is_linked:
            if self.dyn_prop_name:
                layout.prop(self, self.dyn_prop_name)
            else:
                layout.label(t)
        elif self.is_linked:
            layout.label(t + '. ' + SvGetSocketInfo(self))
        else:
            layout.label(t)

    def draw_color(self, context, node):
        return {
            "StringsSocket": (0.6, 1.0, 0.6, 1.0),
            "VerticesSocket": (0.9, 0.6, 0.2, 1.0),
            "MatrixSocket": (.2, .8, .8, 1.0)
        }.get(self.dyn_prop_type, (0.0, 0.0, 0.0, 0.0))
```

**Comment by ly29** [comment]
Something like that could work, of course it is also not so pretty. But less bad than some other options...

**Comment by zeffii** [comment]
I really like the idea that the property used for the slider / bool / whatever is part of the SocketType.

**Comment by zeffii** [comment]
but it is a bit of a joke that there has to be so much work put into making this work on a UI.. whatever way it is approached..

**Comment by ly29** [comment]
It really should be, it makes it clearer where the data comes from and removes the need for bad code. 
All I have lot more appreciation for the standard hierarchy of node sockets now. I really wish changing to such a format would be none breaking but it isn't. Too much cruft going on.

Also the amount of work is because bad decisions earlier, due lack of understanding of the tradeoffs, if would say we don't care about backwards compatibility it becomes so much easier.

**Comment by zeffii** [comment]
that was the purpose of Redux <progress through total gutting breakage>, but perhaps the jump was too radical too soon.

**Comment by ly29** [comment]
A more limited redux just cleaning up nodes and sockets would also be nice. however now we are going way off topic.

**Comment by zeffii** [comment]
agreed..

**Comment by zeffii** [comment]
i think we could tunnel the sliderproperty from the node the socket came from?

```
parent_node_tree:       parent node.inputs (3) (slider from 1)     
--------------------
monad_node_tree:      ng_outputs  (2) <- some_node_input.FloatProperty (1)
```

**Comment by ly29** [comment]
unless we have two instance of a monad node, then that would lead to adjusting one of them changes the other.

**Comment by ly29** [comment]
Imagine something like this for socket.draw()
```python
def draw(self, context, layout, node, text):
    data = bpy.node_groups[node.monad].nodes[self.prop_ref]
    layout.prop(data, "self.prop_name")
```
then changing the exposed property changes all monad nodes interfacing the monad.

**Comment by zeffii** [comment]
maybe stick the dummy nodes in a frame in the monad, and have it hidden by default? this would beat the need to shadow the node tree...  if going that route...

**Comment by ly29** [comment]
something like that yeah

**Comment by zeffii** [comment]
OK.. I'll try with the Frame thing then, but it's not clear to me how that would make monad "instances" easier.

**Comment by zeffii** [comment]
first I want to see if this is a viable way to tunnel a property.. ;)

**Comment by zeffii** [comment]
@ly29 do you know if there's a way to do a `prop_search` on a subset of a collection? like
```
    layout.prop_search(self, "group_name", bpy.data.node_groups, filter('attr', 'sub_group'))
```

**Comment by ly29** [comment]
No, not directly. I guess you can use an enum with a function making the filtering in worst case

**Comment by zeffii** [comment]
Neat! does that work?  will try whenever this frame stuff is done.. ( slowly getting there)

**Comment by zeffii** [comment]
I just realize that Hide in the nodeview doesn't infact hide, but minimizes... kinda diminishes its usefulness..

**Comment by zeffii** [comment]
well, it's displaying the 'tunneled' socket now, but no data is transferred when the parent node's sliders are moved. A problem for tomorrow. ..almost tempted to revert because this just feel ridiculous and brittle (even if the code isn't _too_ ugly.. it's still icky...)

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17347241/05d3aba8-590f-11e6-9e1e-a77be771b603.png)

**Comment by ly29** [comment]
It does feel wrong, doesn't?
For the update to work some changes has to be made, which would make it even more brittle...

**Comment by zeffii** [comment]
am going to revert...

**Comment by ly29** [comment]
Yeah looking at it now it feels like the sane decision.

**Comment by zeffii** [comment]
hope I didn't mess up the revert...

**Comment by zeffii** [comment]
http://stackoverflow.com/a/12049323/1243487

git reset c640aeb   >> will say : Unstaged changes after reset:
git add -A              >> addsthem
git reset --hard       >> bulldoze
git commit -am 'reset to c640aeb'

**Comment by zeffii** [comment]
going to trim some comments from this thread to help it load a bit faster..

**Comment by zeffii** [comment]
```
bpy.ops.node.clipboard_copy()
bpy.ops.node.clipboard_paste()
```
wtf :)

**Comment by zeffii** [comment]
maybe something like this will work... lazy++ first:)
```python
class SvMonadCreateFromSelected(bpy.types.Operator):

    bl_idname = "node.sv_monad_from_selected"
    bl_label = "Create monad from selected nodes (sub graph)"

    group_name = "Monad"

    def execute(self, context):

        ng = context.space_data.edit_tree
        nodes = [n for n in ng.nodes if n.select]
        parent_tree_name = ng.name

        if not nodes:
            self.report({"CANCELLED"}, "No nodes selected")
            return {'CANCELLED'}

        parent_node = ng.nodes.new('SvGroupNodeExp')
        parent_node.location = average_of_selected(nodes)
        bpy.ops.node.clipboard_copy()
        monad = group_make(parent_node, self.group_name)

        bpy.ops.node.sv_switch_layout(layout_name=monad.name)
        
        # by switching, space_data is now different
        path = context.space_data.path
        path.clear()
        path.append(ng[parent_tree_name]) # below the green opacity layer
        path.append(ng[self.group_name])  # top level

        bpy.ops.node.clipboard_paste()

        # switch to parent
        # bpy.ops.node.delete()

        # switch to monad
        # unselect everything?
```

**Comment by ly29** [comment]
If it works great, otherwise my old operator to create a group from selection should work with minimal changes (.bl_idname for nodes etc)

**Comment by ly29** [comment]
https://github.com/nortikin/sverchok/blob/master/utils/group_tools.py

**Comment by zeffii** [comment]
Yeah I did dig through that first before starting this :) I felt there had to be some way to reuse code that's already tried and tested by millions. 

```python
class SvMonadCreateFromSelected(bpy.types.Operator):

    bl_idname = "node.sv_monad_from_selected"
    bl_label = "Create monad from selected nodes (sub graph)"

    group_name = StringProperty(default="Monad")

    def execute(self, context):

        ng = context.space_data.edit_tree
        nodes = [n for n in ng.nodes if n.select]

        if not nodes:
            self.report({"CANCELLED"}, "No nodes selected")
            return {'CANCELLED'}

        parent_node = ng.nodes.new('SvGroupNodeExp')
        parent_node.select = False
        parent_tree = parent_node.id_data
        parent_node.location = average_of_selected(nodes)
        bpy.ops.node.clipboard_copy()

        monad = group_make(parent_node, self.group_name)
        bpy.ops.node.sv_switch_layout(layout_name=monad.name)
        
        # by switching, space_data is now different
        path = context.space_data.path
        path.clear()
        path.append(parent_tree) # below the green opacity layer
        path.append(monad)  # top level

        bpy.ops.node.clipboard_paste()

        # remove nodes from parent_tree
        for n in reversed(nodes):
            parent_tree.nodes.remove(n)

        return {'FINISHED'}
```

**Comment by zeffii** [comment]
it might be sane to have a filter list to unselect certain kinds of nodes if they are in the user's selection... but that's the easy bit.

**Comment by ly29** [comment]
that code does even spend that much time of copying, just collecting links and such things is quite enough of a chore.

**Comment by zeffii** [comment]
It probably makes sense to place the monad input and output at the extremes/bounding box of the copied nodes, I find myself moving them all the time..

**Comment by zeffii** [comment]
```text
              _____________________________________
             |                                     |
             | [_____]                             | 
             |                                     | 
[ outs    ]  |         [_____]             [_____] |  [ ins ]
             |                                     |             
             |         [_____]                     |             
             |                             [_____] |
             |_____________________________________|
```

**Comment by zeffii** [comment]
sv_monad_tools.py  holds the Operators and the utils now..

**Comment by zeffii** [comment]
![image](https://cloud.githubusercontent.com/assets/619340/17381053/61de9410-59ca-11e6-8ca9-77b587d6a5de.png)

[now monad IO nodes are placed](https://github.com/nortikin/sverchok/blob/38986f633fe601de42ce18cac39ee069d1f870c8/utils/sv_monad_tools.py#L43-L53) at  `(min_y + max_y) / 2`  and `min_x` and `min_y` respectively.

**Comment by zeffii** [comment]
I realize now the one important difference between my simpler `make_monad_from_selected` vs your original counterpart is that you auto connect the head and tail of the selected nodes to the monad IO nodes, and the parent node.. This means my implementation is only partially successful at the moment

**Comment by zeffii** [comment]
also.. perhaps monad IO should auto recast their Socket type, and also auto rename in that case, but not auto rename when the socket type doesn't change

**Comment by zeffii** [comment]
`get_relinks()` and `relink()` are a little broken.. but committed anyway.

**Comment by zeffii** [comment]
let's add more naming convention to make it easier to talk about.

- _pre monad nodes_: all selected nodes prior to turning them into a monad.
- _periphery nodes_ : are those nodes which are not part of the pre_monad set but that do have inputs or outputs attached to one of more of the pre_monad set.

after monad creation 
- links should be re-established between the monad's parent node and the periphery nodes, 
- internally the monad links should connect those links to the monad's input and outputs.

**Comment by zeffii** [comment]
I may copy liberally from `group_tools.py/SvNodeGroupCreator` :)

**Comment by ly29** [comment]
Feel free to do it.
Also an important point is that one socket may have many to sockets. Or also why ```socket.links``` is an array. (well it is a property function actually but that is a different matter)

**Comment by zeffii** [comment]
ah yes, `outputs[index].links[:]`  nodes can output to various other nodes from one socket, but can only receive (or we have limited it to do so) into `inputs[index].links[0]`.

..this way take a while to figure out a nice way to do it, no ETA.

**Comment by ly29** [comment]
The ugly way is in ```SvNodeGroup```

**Comment by zeffii** [comment]
I'm under no illusion that i'll come up with anything nicer than SvNodeGroup, but i still want to have a think about it

**Comment by zeffii** [comment]
Possible the code I just pushed will suffice as 'make monad from selection' - else i'll need to really just rip your code.. because I have little patience for this.

**Comment by zeffii** [comment]
I think maybe the monad inputs could be arrange by node.location.y so as to avoid crossing noodles?

![image](https://cloud.githubusercontent.com/assets/619340/17446266/b45cf0f4-5b48-11e6-9a4b-b5a1e2ef54b6.png)

**Comment by ly29** [comment]
A very sane thought, I can do some testing tomorrow. 

Some sanity testing might be in order.

**Comment by zeffii** [comment]
`SvMonadCreateFromSelected(Operator) `  would be good idea to offer a mode which does not auto connect peripherals, simple bypass for `get_relinks()` and `relink()`.

**Comment by zeffii** [comment]
a nitpick ..
not sure what the process is for removing menu items from NODE_MT_node and inserting our own implementation in its place. Is it even possible? ..I really don't like the current situation where there are  items in that menu that don't apply to Sverchok...

**Comment by zeffii** [comment]
maybe `Ctrl+G` to group and `Ctrl+Shift+G` to group without peripheral relinks.

yay / nay ?

**Comment by ly29** [comment]
yay from me, how do does overriding and operator for certain node type look like btw?
git did something messed up (or I did with git actually).

**Comment by zeffii** [comment]
I remember a long time ago i / we asked LukasT about overriding the Add New operator for nodeview, so as to attach a `fake_user` to each new tree, but we never implemented it. Suggest to me that we either didn't understand it or the method was not so pretty..

re git:  last time you did a git push on this branch I hadn't pulled before my latest commit,. In order to save time I deleted the local sverchok and cloned the repo again to get your changes.

**Comment by zeffii** [comment]
overloading the operator: https://github.com/nortikin/sverchok/issues/356#issuecomment-50598373

**Comment by ly29** [comment]
Yeah I messed up git, really sorry about that.

**Comment by zeffii** [comment]
.. i will revert to 185e5ff  and push

**Comment by zeffii** [comment]
pushed.

**Comment by zeffii** [comment]
i split the code into a few files
```
ui/nodeview_keymaps.py
ui/sv_monad_panel.py
utils/sv_monad_tools.py
```

**Comment by ly29** [comment]
I give up, I will stop messing around with git.

**Comment by zeffii** [comment]
uchh i don't think i restored to the last commit..

**Comment by ly29** [comment]
I think I messed it up again, will  remove my local copy and reclone when it is restored

**Comment by zeffii** [comment]
ok :)

**Comment by zeffii** [comment]
give me about 10 minutes..

**Comment by zeffii** [comment]
I should have restored to : b6f443e

**Comment by zeffii** [comment]
should be ok now.  `git pull --rebase` if you need to

**Comment by ly29** [comment]
Okay, now are talking.

- Monad in monad, I think this should be allowed but how editing works out is worth thinking about
- if you make a monad with one node, ```propose_io_location``` fails.

```python
    min_x, max_x = min(*x_locs), max(*x_locs)
```
Could be changed to
```python
    min_x, max_x = min(x_locs), max(x_locs)
```
The expansion of arguments isn't need anyway.

**Comment by ly29** [comment]
Will make  branch to experiment a bit with creating a dynamic class representing each node group.

**Comment by zeffii** [comment]
oh right -- will drop the expansion.. done 4ad5215

Monad in Monad could definitely be cool, and may be less complicated to implement than it seems.

**Comment by ly29** [comment]
It should just work, some assumptions might have to be dropped.

```min``` and ```max``` are clever:

```python
    x_locs = [n.location[0] for n in nodes]
    y_locs = [n.location[1] for n in nodes]
    min_x, max_x = min(*x_locs), max(*x_locs)
    min_y, max_y = min(*y_locs), max(*y_locs)
```
Could be

```python
    locs = [n.location[:] for n in nodes]
    min_x, min_y = min(locs)
    max_x, max_y = max(locs)
```

**Comment by zeffii** [comment]
nice. i like that shorter approach

**Comment by zeffii** [comment]
(but it produces different results..)

**Comment by zeffii** [comment]
```python
    x_locs = [n.location[0] for n in nodes]
    y_locs = [n.location[1] for n in nodes]
    min_x, max_x = min(x_locs), max(x_locs)
    min_y, max_y = min(y_locs), max(y_locs)
```
![image](https://cloud.githubusercontent.com/assets/619340/17484052/a4e294c0-5d88-11e6-8509-caa9f9f221a1.png)

vs
```python
    locs = [n.location[:] for n in nodes]
    min_x, min_y = min(locs)
    max_x, max_y = max(locs)
```


![image](https://cloud.githubusercontent.com/assets/619340/17484083/bea1b990-5d88-11e6-9547-e7c1010013a9.png)

**Comment by zeffii** [comment]
because: 
```
>>> max([[20,30],[40,50],[70,80]])
[70, 80]
```

**Comment by ly29** [comment]
Okay, I have to be more careful. I was sure there was clever way to do like that.

Anyway, onwards to creating dynamic node groups with properties. 
First commits up but not functional yet.
See [dynamic_monad_node](https://github.com/nortikin/sverchok/tree/dynamic_monad_node)

Doing this some assumptions change so everything that uses ```parent_node``` is also being reworked.

**Comment by ly29** [comment]
Some progress
![skarmavbild 2016-08-09 kl 08 57 51](https://cloud.githubusercontent.com/assets/6241382/17507409/caacf77c-5e0f-11e6-9263-8145f49ecab5.png)
Before making group
![skarmavbild 2016-08-09 kl 09 00 20](https://cloud.githubusercontent.com/assets/6241382/17507408/caa54f36-5e0f-11e6-9415-05e8127372a7.png)
After making group and unlinking.

Still many issues.

**Comment by zeffii** [comment]
that's what we want :)   code-wise I must confess I don't understand much of it yet ( yet, of course, none of the syntax is new and i am comfortable will dynamic class creation). I will put time in today to get working knowledge of what you're doing.

**Comment by ly29** [comment]
I going to of bit of sorting out dependencies hell, the problem with F8 development.

**Comment by zeffii** [comment]
https://github.com/nortikin/sverchok/compare/b6f443e...1bea57f   nice.

Is some dependecy hell is my fault? ..because i'm not registering the classes inside sv_monad_tools, but rather importing them inside group_exp.py then registering, and there's a few more ..

**Comment by zeffii** [comment]
or (what i expect..) the registering / unregistering of dynamic classes...  ?

**Comment by ly29** [comment]
Yes, kind of, where things belong needs to be looked over a little bit. But also I am slithy rusty with the subtler point of python.

**Comment by ly29** [comment]
The the dynamic class thing will come with it's own issues. 

What works now:

- Creating a node group with relinks and getting sockets with properties
- processing of the node group

What does not work
- Further editing
- Or anything else basically.

What needs to be done:
- Updating a node group should trigger ```make_class_from_monad``` and update all instance of the class collecting links etc.
- The needs to be a sanity check to handle more than one property with the same name
- Update all the the tools/operators to use this interface

**Comment by zeffii** [comment]
> The needs to be a sanity check to handle more than one property with the same name
first come first serve, if two 'x'  then 'x', 'x2'. if user renames manually to an identical name, then the other socket should receive the postfix (and this to cascade to the parent node).

seems like you are moving towards Lukastoenne's reference imlementation, i'm OK with that.

**Comment by zeffii** [comment]
let me get this straight: each instance of a monad is a unique class registered whenever we ctrl+<shift>+G. ?

also... i dont understand the appending to the socket here of `data`
```python
+    in_socket = []
+
+    for socket in monad_inputs.outputs:
+        if socket.is_linked:
+            other = get_other_socket(socket)
+            prop_name = getattr(other, "prop_name")
+            if prop_name:
+                cls_dict[prop_name] = getattr(other.node.rna_type, prop_name)
+            data = [socket.name, socket.bl_idname, prop_name if prop_name else None]
+            in_socket.append(data)
+
+    out_socket = []
+    for socket in monad_outputs.inputs:
+        if socket.is_linked:
+            data = [socket.name, socket.bl_idname]
+            out_socket.append(data)
```

**Comment by ly29** [comment]
For each monad there is unique monad node class with the corresponding properties to show in the sockets. This monad node can have many instances.


The data (bad name) is collected into ```input_template``` and ```output_template```, it is the for each socket data needed to create the socket.
```
>>> D.node_groups['NodeTree'].nodes['Group Exp'].input_template
[['vectors', 'VerticesSocket', None], ['Angle', 'StringsSocket', None], ['vertices', 'VerticesSocket', None], ['Rotation', 'VerticesSocket', None]]
```

```python
    def sv_init(self, context):
        self.use_custom_color = True
        self.color = MONAD_COLOR

        for socket_name, socket_bl_idname, prop_name in self.input_template:
            s = self.inputs.new(socket_bl_idname, socket_name)
            if prop_name:
                s.prop_name = prop_name
        for socket_name, socket_bl_idname in self.output_template:
            self.outputs.new(socket_bl_idname, socket_name)
```

**Comment by zeffii** [comment]
ok,  monad_inputs  are the input socket on (what i was calling) 'parent_node' ?

**Comment by zeffii** [comment]
or an `interim structure` (*put_template)  used to populate the input sockets on the parent_node

**Comment by ly29** [comment]
```monad_inputs``` are taken from the ```SvGroupInputsNodeExp``` in the monad.

There are also some changes in the order things happen in the monad creation.

**Comment by ly29** [comment]
I just made the changes needed to make it work a little bit.
For example every edit operation, renaming etc there needs to be some mechanism to update all

**Comment by ly29** [comment]
![skarmavbild 2016-08-09 kl 10 57 31](https://cloud.githubusercontent.com/assets/6241382/17510664/2f89f0a4-5e20-11e6-8bbe-e2551f7bc169.png)
Node instances of a monad.

**Comment by zeffii** [comment]
looking good! 

a side node: hey maybe we can make the redux run inside a monad (experimentally) eventually. could help bootstrap / transition.

**Comment by ly29** [comment]
Actually for most cases of editing, changing the templates and updating sockets could be done without registering a new class. Just changing the ```input_template``` etc and update as needed.

**Comment by ly29** [comment]
Looking at the operator code it looks reasonably simple to extend to the fact that there might be more than one node to edit.

Aslong as each edit is atomic and only affects one thing at a time it might be easier to something like below. 
```python
    def execute(self, context):
        node, kind, socket = get_data(self, context)
        IO_node_sockets = getattr(node, kind)
        pos = self.pos

        if self.direction == 0:
            IO_node_sockets.remove(socket)     # I/O interface (subgroup)
            # parent_sockets.remove(parent_sockets[pos])
            for node in node.id_data.node_instances:
                  sockets = get_sockets(node, kind)
                  sockets.remove(sockets[pos])
        else:
            def wrap_around(current_idx, direction, member_count):
                return (current_idx + direction) % member_count

            IO_new_pos = wrap_around(pos, self.direction, len(IO_node_sockets)-1)
            IO_node_sockets.move(pos, IO_new_pos)

            #parent_new_pos = wrap_around(pos, self.direction, len(parent_sockets))
           # parent_sockets.move(pos, parent_new_pos)
           for node in node.id_data.node_instances:
                sockets = get_sockets(node, kind)
                new_pos = wrap_around(pos, self.direction, len(sockets))
                sockets.move(pos, new_pos)
      # then make sure to update monad to add new things
      make_class_from_monad(node.id_data.name)

```

**Comment by zeffii** [comment]
how to make a clear distinction between
- monads instances (edit one edit all ) 
    - or lock instances and only let edits take place in the 'master' monad.
- and unique duplicates (no edit propagation

**Comment by ly29** [comment]
I think the sane approach is to consider them like resuable blocks of code. 
So if you edit one you edit all instances of it, if you want to duplicate you need to explicitly create a copy. 
Click the edit button brings up the same monad. 

Could even be that you have to expand the monad and create a new monad. 
A minimal interface to achieve this would suggest that approach.

**Comment by zeffii** [comment]
ok, so if you are in an instance we display a large button in the UI to 'Make Unique'.

**Comment by nortikin** [comment]
OMG.

**Comment by zeffii** [comment]
@ly29 perhaps start a new thread regarding the move towards dynamic nodes, and copy the todolist and untick those items which are not yet implemented/working in `dynamic_monad_node`, 

this thread got too long already :) slow for browsing new stuff

**Comment by zeffii** [comment]
close for now.
